### PR TITLE
Upgrade socket2 dependency

### PR DESCRIPTION
**Pull request by faern**
Upgrades to a version not making invalid assumptions about the memory layout of `std::net::SocketAddr`.

Helps unblock https://github.com/rust-lang/rust/pull/78802. See that PR for more details.

### CODE REVIEW

**Review by bluejekyll — APPROVED** [review]
LGTM

**Comment by bluejekyll** [comment]
Thanks for the explanation, will merge once tests are passing.

**Comment by faern** [comment]
Thank you. Does it make sense to backport to the 0.19 branch as well maybe? I see it's still seeing releases.

**Comment by bluejekyll** [comment]
I don't think that's necessary. It's a minor release, so will be updated if folks do a cargo update. I would think if it's a big concern, we should add it to the rustsec db: https://rustsec.org/

Then folks will upgrade as necessary.

**Comment by djc** [comment]
I'm inclined to think that it would be good to make a 0.19 release including this, since the 0.20 release will be blocked on tokio 0.3 as well, and it would be good to allow the ecosystem to move away from the dependency on socket layout without doing incompatible upgrades where possible.

@faern thanks for your work on this issue to help move the ecosystem forward!

**Comment by faern** [comment]
If `0.20` is going to be using `tokio 0.3` I also really think we should backport this to `0.19` as many will be stuck in that transition for quite a while.

> I would think if it's a big concern, we should add it to the rustsec db: https://rustsec.org/

That is a really good idea. As the main change moves forward we probably should! Since staying on the older versions will cause UB after the standard library changes. But I won't report that until I have made the majority of the ecosystem automatically upgrade thanks to time (and doing pushes like this to popular crates)

**Pull request merged**
